### PR TITLE
Make various primitives `non_null_value`

### PR DESCRIPTION
**Pull request by dkalinichenko-js**
Default all immediates to non-null.

Make all pre-defined types non-null, temporarily with the exception of `lazy_t`.

Make extensible variants, boxed variants, boxed records, tuples, polymorphic variants, functions and first-class modules non-null.
This excludes single-field records and variants tagged with `[@@unboxed]`, which inherit the layout of the field.

Add tests for those types. Add tests for type variables in modules and functions. Add tests for stdlib definitions.

This PR does not implement "hiding" of `non_null_value` when `layouts_alpha` is disabled, so it shouldn't be (?) merged
into the compiler by itself.

### CODE REVIEW

File: ocaml/typing/jkind.mli [hunk 1]
```diff
@@ -151,6 +151,8 @@ val void : why:void_creation_reason -> t
 (** This is the jkind of normal ocaml values *)
 val value : why:value_creation_reason -> t
 
+val non_null_value : why:non_null_value_creation_reason -> t
```
Comment (goldfirere) [inline ocaml/typing/jkind.mli:154]: Please add a comment, like the others.

File: ocaml/typing/predef.ml [hunk 2]
```diff
@@ -321,7 +321,7 @@ let build_initial_env add_type add_extension empty_env =
       ~jkind_annotation:Immediate
```
Comment (goldfirere) [inline ocaml/typing/predef.ml:321]: Can't comment on the line, but `ident_bool` should be non-null immediate.

Comment (goldfirere) [inline ocaml/typing/predef.ml:321]: Should be non-null immediate.

File: ocaml/typing/predef.ml [hunk 3]
```diff
@@ -321,7 +321,7 @@ let build_initial_env add_type add_extension empty_env =
...
   |> add_type ident_exn
        ~kind:Type_open
-       ~jkind:(Jkind.value ~why:Extensible_variant)
+       ~jkind:(Jkind.non_null_value ~why:Extensible_variant)
   |> add_type ident_extension_constructor
   |> add_type ident_float
   |> add_type ident_floatarray
```
Comment (goldfirere) [inline ocaml/typing/predef.ml:327]: Actually everything that's immediate here should be non-null immediate.

**Review by goldfirere — COMMENTED** [review]
Thorough testing -- thanks.

Though I suppose it could be done in another PR, we should sort out the story around non-null immediates. And I'd generally expect `value_creation_reason` to be pretty empty by the time this patch is done -- hollowing that out could be a goal to chase.

File: ocaml/typing/predef.ml (hunk 3 above)
Comment (dkalinichenko-js) [reply ocaml/typing/predef.ml:327]: This PR makes every `immediate` non-nullable. Instead, we can't express a nullable immediate.

File: ocaml/testsuite/tests/typing-layouts-non-null-value/arguments.ml [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  val mk : 'a -> 'a t
+end = struct
+  type 'a t = 'a
+
+  let mk x = x
+end
+
+(* By default, type parameters to abstract types are nullable: *)
+
+let _ = M1.mk (Fake_or_null.some 5)
```
[the remark this answers is not in the document]
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-non-null-value/arguments.ml:77]: Will be addressed in future PRs.

File: ocaml/testsuite/tests/typing-layouts-non-null-value/stdlib_defs.ml [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+let _ = Atomic.make (Possibly_null.create 2)
+;;
+
+[%%expect{|
+- : Possibly_null.t Atomic.t = <abstr>
+|}]
+
+(* CR layouts v3.0: [Bigarray.t] should be non-null. *)
+
+let _ = id_non_null_value (Bigarray.Array1.create Bigarray.float64 Bigarray.c_layout 3)
```
[the remark this answers is not in the document]
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-non-null-value/stdlib_defs.ml:63]: Since those tests are in `typing-layouts-non-null-value`, I think we should make another directory for testing actual null values.

File: ocaml/testsuite/tests/typing-layouts/basics.ml [hunk 6]
```diff
@@ -2383,8 +2383,8 @@ Line 2, characters 0-14:
 2 | and 'a t2 = 'a
     ^^^^^^^^^^^^^^
 Error:
-       The layout of 'a t2 is value, because
-         it instantiates an unannotated type parameter of t2, defaulted to layout value.
+       The layout of 'a t2 is '_representable_layout_9, because
```
[the remark this answers is not in the document]
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-layouts/basics.ml:2386]: This is fine for now. But it's another example of why we need kind variables, not sort variables.

File: ocaml/testsuite/tests/typing-layouts-non-null-value/arguments.ml (hunk 4 above)
[the remark this answers is not in the document]
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-non-null-value/arguments.ml:93]: Added a test + updated the comment to say it should default to non-null.

File: ocaml/testsuite/tests/typing-layouts-non-null-value/arguments.ml [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+let _ = [ Fake_or_null.none ]
+
+let _ = [| Fake_or_null.some 3 |]
+
+let _ = [: Fake_or_null.some "test " :]
+
+let _ = Some (Fake_or_null.some 4.2)
+
+let _ = lazy (Fake_or_null.none)
```
[the remark this answers is not in the document]
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-non-null-value/arguments.ml:54]: Tested in `basics.ml`, this tests arguments.

File: ocaml/typing/jkind.mli (hunk 1 above)
Comment (dkalinichenko-js) [reply ocaml/typing/jkind.mli:155]: Done.

File: ocaml/testsuite/tests/typing-layouts-non-null-value/stdlib_defs.ml [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+                              ^^^^^^^^^^^^^^^^^^
+Error: This expression has type ('a, 'b) Hashtbl.t
+       but an expression was expected of type ('c : non_null_value)
+       The layout of ('a, 'b) Hashtbl.t is value, because
+         of layout requirements from an imported definition.
+       But the layout of ('a, 'b) Hashtbl.t must be a sublayout of non_null_value, because
+         of the definition of id_non_null_value at line 1, characters 4-21.
+|}]
+
+let _ = Hashtbl.add (Hashtbl.create 1) "test" (Possibly_null.create 7)
```
[the remark this answers is not in the document]
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-non-null-value/stdlib_defs.ml:132]: Will be tested in a different group of tests when we actually have null pointers.

**Review by goldfirere — APPROVED** [review]

**Comment by goldfirere** [comment]
In light of the inference challenges we're seeing, is this still good to merge? If so, go for it.

**Comment by dkalinichenko-js** [comment]
This should work even if we can't figure out inference for abstract types.
